### PR TITLE
feat: Add toHaveValue matcher #82

### PR DESCRIPTION
**Pull request by lukaszfiszer**
**What**:
Add `toHaveValue` matcher

**Why**:
As discussed in #82 

**How**:
Reused a lot of existing logic from `toHaveFormValue` and moved some internal functions  to `utils.js` so that they can be shared between both modules.

I had a lot of thoughts on how to implement this matcher for `<input type="checkbox" />` and `<input type="radio" />`, because those elements value can be different depending on other elements from the same group (sharing the same `name`). Finally I've decided to throw an error when a checkbox or radio input is passed, suggesting users to use `toHaveFormValue` instead. In the future, I think I would most make sense to implement `toBeChecked` matcher - a perfect pair with to `toHaveValue` for single form elements. 

**Checklist**:

* [x] Documentation
* [x] Tests
* [x] Updated Type Definitions
* [x] Ready to be merged 
* [x] Added myself to contributors table

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -687,6 +688,63 @@ expect(element).not.toHaveTextContent('content')
...
+### `toHaveValue`
+
+```typescript
+toHaveValue(value: string | string[])
+```
+
+This allows you to check whether the given form element has the specified value.
+It accepts `<input>`, `<select>` and `<textarea>` elements with the exception of
+of `<input type="checkbox">` and `<input type="radio">`, which can be meaningfully
+matched only using [`toHaveFromValue`](#tohaveformvalues).
```
Comment (gnapse) [inline README.md]: Typo `toHaveFromValue` should be `toHaveFormValue`

File: README.md [hunk 2]
```diff
@@ -687,6 +688,63 @@ expect(element).not.toHaveTextContent('content')
...
+
+For all other form elements, the value is matched using the same algorithm
+as in [`toHaveFromValue`](#tohaveformvalues) does.
```
Comment (gnapse) [inline README.md]: Same typo.

**Comment by lukaszfiszer** [comment]
@gnapse I have fix the typos.

Another idea that I had for checkbox and radio inputs was to allow the user to pass a parent element of those elements:

```html
<div data-testid="contact">
   <label><input type="radio" name="contact" value="email">Email</label>
   <label><input type="radio" name="contact" value="phone">Phone</label>
   <label><input type="radio" name="contact" value="post">Post</label>
</div>
```

```js
expect(getByTestId('contact')).toHaveValue('phone')
```

But then probably you could  just use `toHaveFormValues` for that.

Let me know what you think

**Comment by kentcdodds** [comment]
What do we need to do to help get this across the finish line?

**Comment by gnapse** [comment]
I probably forgot to keep track of this and approve or request changes at the time. But now it seems there are conflicts, and also the build is failing for some reason (could it have something to do with the change of url of the github repo when moved to an org @kentcdodds?  When I visit the travis details link above I get "We couldn't find the repository gnapse/jest-dom").

@lukaszfiszer sorry about not being on top of this. Can you please fix the conflicts? We'll figure out the travis issue in the mean time.

**Comment by lukaszfiszer** [comment]
@gnapse No worries. I will fix the conflicts over the weekend.

**Comment by lukaszfiszer** [comment]
@gnapse rebased against master and fixed the broken test. ready to merge

File: extend-expect.d.ts [hunk 3]
```diff
@@ -20,5 +20,6 @@ declare namespace jest {
       text: string | RegExp,
       options?: {normalizeWhitespace: boolean},
     ): R
+    toHaveValue(value?: string | string[]): R
```
Comment (jgoz) [inline extend-expect.d.ts]: Should this be `value?: string | string[] | number`?

File: README.md [hunk 4]
```diff
@@ -690,6 +691,63 @@ expect(element).not.toHaveTextContent('content')
 
 <hr />
 
+### `toHaveValue`
+
+```typescript
+toHaveValue(value: string | string[])
```
Comment (jgoz) [inline README.md]: And should this one be `value?: string | string[] | number`?

File: src/__tests__/to-have-value.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+describe('.toHaveValue', () => {
+  test('handles value of text input', () => {
+    const {queryByTestId} = render(`
+        <input type="text" value="foo" data-testid="with" />
+        <input type="text" value=""  data-testid="empty" />
+        <input type="text" data-testid="without" />
+    `)
+
+    expect(queryByTestId('with')).toHaveValue('foo')
+    expect(queryByTestId('with')).not.toHaveValue('bar')
```
Comment (jgoz) [inline src/__tests__/to-have-value.js]: Would it make sense to add a test case like this?

```js
expect(queryByTestId('with')).toHaveValue();
```

The corresponding `.not.toHaveValue()` case is already there for `empty`.

**Comment by lukaszfiszer** [comment]
@jgoz Thanks for review. I've fixed the typings, added additional test cases and fixed the behaviour of `.toHaveValue()` (without argument) - it asserts if input/select has *any* value.

File: src/index.js [hunk 6]
```diff
@@ -29,7 +30,9 @@ export {
   toBeVisible,
   toBeDisabled,
   toBeEnabled,
+  toHaveValue,
   toBeRequired,
   toBeInvalid,
   toBeValid,
+  toHaveValue
```
Comment (jgoz) [inline src/index.js]: This is being exported twice

File: src/to-have-value.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+export function toHaveValue(htmlElement, expectedValue) {
+  checkHtmlElement(htmlElement, toHaveValue, this)
+
+  if (
+    htmlElement.tagName.toLowerCase() === 'input' &&
+    ['checkbox', 'radio'].includes(htmlElement.type)
+  ) {
+    throw new Error(
+      'input with type=checkbox or type=radio cannot be used with .toHaveValue(). Use .toHaveFormValue() instead',
```
Comment (jgoz) [inline src/to-have-value.js]: Minor: `.toHaveFormValue()` should be `.toHaveFormValues()`

**Comment by jgoz** [comment]
@lukaszfiszer sorry for the delayed response, I missed the notification on this. Just two very minor issues that I can see, one of which is causing the CI failure.

**Comment by lukaszfiszer** [comment]
@jgoz fixed issues and squash-rebased against master. Hope to get it merged!

File: README.md [hunk 8]
```diff
@@ -839,6 +840,63 @@ expect(element).not.toHaveTextContent('content')
...
+expect(emptyInput).not.toHaveValue()
+expect(selectInput).not.toHaveValue(['second', 'third'])
+```
+
+##### Using dom-testing-library
+
+```javascript
+const {getByTestId} = render(/* Rendered HTML */)
+
+const textInput = getByTestId('[data-testid="input-text"]')
```
Comment (kentcdodds) [inline README.md]: I think these should be: `getByTestId('input-text')` and so-forth.

Comment (lukaszfiszer) [reply README.md]: Good catch! Fixed.

**Pull request merged**